File: crates/mediaway-container/src/mp3.rs
//! Mediaway-typed MPEG audio (Layer III / "MP3") mux + demux over [`mpeg_audio`].
//!
//! Like ADTS, MPEG audio carries no timing metadata of its own —
//! [`Demuxer::poll_packet`] synthesizes `pts`/`duration` from a running sample
//! count using the standard per-version frame size (1152 samples for MPEG-1
//! Layer III, 576 for MPEG-2/2.5 Layer III — half-rate).
//!
//! [`Muxer`] does not implement [`crate::Mux`]: real Layer III streams flip a
//! per-frame padding bit to average out fractional frame lengths (a
//! bit-reservoir detail `mpeg_audio::Muxer::write_frame` requires as an
//! explicit argument), which the generic `Packet`-based trait has no slot
//! for — exposing the real, correct API is more honest than forcing a fit
//! that would silently default padding away.

#![forbid(unsafe_code)]

use crate::Demux;
use mediaway_common::{Bytes, CodecKind, Packet, Rational, StreamInfo};
use mpeg_audio::{Demuxer as CoreDemuxer, FrameHeader, MpegVersion, Muxer as CoreMuxer};

/// MP3 mux/demux error (same as [`mpeg_audio::Error`]).
pub type Error = mpeg_audio::Error;

/// Samples per Layer III frame — 1152 for MPEG-1, 576 (half-rate) for MPEG-2/2.5.
#[must_use]
pub const fn samples_per_frame(version: MpegVersion) -> u64 {
    match version {
        MpegVersion::Mpeg1 => 1152,
        // Covers MPEG-2/2.5 (half-rate) and any future `MpegVersion` variant.
        _ => 576,
    }
}

/// Live mux session for a fixed [`FrameHeader`].
///
/// Bitrate/sample-rate/channel mode stay constant for the session's
/// lifetime, matching real Layer III streams this crate targets — VBR would
/// need a new header per frame, out of scope.
#[derive(Debug)]
pub struct Muxer {
    inner: CoreMuxer,
}

impl Muxer {
    /// Open a mux session for `header`.
    ///
    /// # Errors
    ///
    /// Returns an error for a non-standard bitrate/sample-rate combination.
    pub fn new(header: FrameHeader) -> Result<Self, Error> {
        Ok(Self {
            inner: CoreMuxer::new(header)?,
        })
    }

    /// Append one already-encoded Layer III frame body.
    ///
    /// # Errors
    ///
    /// Returns an error when `frame_body`'s length doesn't match what the
    /// header's bitrate/sample-rate/`padding` combination requires.
    pub fn write_frame(
        &self,
        frame_body: &[u8],
        padding: bool,
        out: &mut Vec<u8>,
    ) -> Result<(), Error> {
        self.inner.write_frame(frame_body, padding, out)
    }
}

/// Demuxer wrapping [`mpeg_audio::Demuxer`] with a Mediaway stream cache.
#[derive(Debug, Default)]
pub struct Demuxer {
    inner: CoreDemuxer,
    streams: Vec<StreamInfo>,
    samples_seen: u64,
}

impl Demuxer {
    /// Empty demuxer.
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Feed bytes.
    pub fn push_bytes(&mut self, chunk: &[u8]) {
        self.inner.push_bytes(chunk);
    }

    /// The single implicit stream, once the first frame's header has been
    /// parsed (empty before that).
    #[must_use]
    pub fn streams(&self) -> &[StreamInfo] {
        &self.streams
    }

    /// Next demuxed packet, synthesizing `pts`/`duration` from a running
    /// samples-per-frame count (see module docs).
    pub fn poll_packet(&mut self) -> Option<Packet> {
        let payload = self.inner.poll_frame().ok().flatten()?;
        let header = self.inner.header()?;
        if self.streams.is_empty() {
            let channels = if matches!(header.channel_mode, mpeg_audio::ChannelMode::Mono) {
                1
            } else {
                2
            };
            self.streams.push(StreamInfo::Audio {
                id: 0,
                codec: CodecKind::Mp3,
                time_base: Rational::new(1, header.sample_rate),
                extra_data: Bytes::new(),
                sample_rate: header.sample_rate,
                channels,
            });
        }
        let frame_samples = samples_per_frame(header.version);
        let pts = i64::try_from(self.samples_seen).unwrap_or(i64::MAX);
        self.samples_seen += frame_samples;
        Some(Packet {
            stream_id: 0,
            pts,
            dts: pts,
            duration: frame_samples,
            is_keyframe: true,
            is_discard: false,
            payload,
        })
    }
}

#[allow(clippy::use_self)]
impl Demux for Demuxer {
    fn push_bytes(&mut self, chunk: &[u8]) {
        Demuxer::push_bytes(self, chunk);
    }

    fn streams(&self) -> &[StreamInfo] {
        Demuxer::streams(self)
    }

    fn poll_packet(&mut self) -> Option<Packet> {
        Demuxer::poll_packet(self)
    }
}

#[cfg(test)]
#[path = "mp3_tests.rs"]
mod tests;

File: bindings/nodejs/packages/container/src/mp3.ts
/**
 * @mediaway/container — MP3 (MPEG Layer III) mux + demux (adr/container/0007-mp3-c-abi.md).
 *
 * A fixed header for the mux session's lifetime (no track registration at
 * all); `writeFrame` takes an explicit `padding` bit no `Packet` has a slot for.
 */

import { container, copyBytes, type RawMp3FrameHeader, type RawPacket, type RawStreamInfo } from "@mediaway/ffi";
import { MediawayError, check, type Packet } from "./index.js";

export type MpegVersion = "mpeg1" | "mpeg2" | "mpeg2_5";
export type ChannelMode = "stereo" | "joint_stereo" | "dual_channel" | "mono";

const MPEG_VERSION_TO_ABI: Record<MpegVersion, number> = { mpeg1: 0, mpeg2: 1, mpeg2_5: 2 };
const CHANNEL_MODE_TO_ABI: Record<ChannelMode, number> = {
  stereo: 0,
  joint_stereo: 1,
  dual_channel: 2,
  mono: 3,
};

/** Fixed Layer III frame header for `Mp3Muxer` — bitrate/sample rate/
 * channel mode stay constant for the whole mux session's lifetime. */
export interface Mp3FrameHeader {
  version: MpegVersion;
  /** Must be one of the 14 standard Layer III rates for `version`. */
  bitrateKbps: number;
  /** Must be one of the 3 standard rates for `version`. */
  sampleRate: number;
  channelMode: ChannelMode;
}

export interface Mp3StreamInfo {
  sampleRate: number;
  channels: number;
  extraData: Buffer;
}

export class Mp3Muxer {
  private handle: unknown;

  /** `header` must be a standard Layer III bitrate/sample-rate combination for its version. */
  constructor(header: Mp3FrameHeader) {
    const raw: RawMp3FrameHeader = {
      version: MPEG_VERSION_TO_ABI[header.version],
      bitrate_kbps: header.bitrateKbps,
      sample_rate: header.sampleRate,
      channel_mode: CHANNEL_MODE_TO_ABI[header.channelMode],
    };
    this.handle = container.mp3MuxerCreate(raw);
    if (!this.handle) {
      throw new MediawayError(1, "non-standard MP3 bitrate/sample-rate combination, or the native call panicked");
    }
  }

  /** Append one already-encoded Layer III frame body. Fails with
   * INVALID_PACKET when `frameBody`'s length doesn't match what the
   * header's bitrate/sample-rate/padding combination requires. */
  writeFrame(frameBody: Buffer, padding: boolean): Buffer {
    const outData: [unknown] = [null];
    const outLen: [number] = [0];
    check(container.mp3MuxerWriteFrame(this.handle, frameBody, frameBody.length, padding, outData, outLen));
    const data = copyBytes(outData[0], outLen[0]);
    if (outLen[0] > 0) container.bufferFree(outData[0], outLen[0]);
    return data;
  }

  close(): void {
    if (this.handle) {
      container.mp3MuxerClose(this.handle);
      this.handle = null;
    }
  }
}

export class Mp3Demuxer {
  private handle: unknown;

  constructor() {
    this.handle = container.mp3DemuxerCreate();
    if (!this.handle) throw new MediawayError(7, "MP3 demuxer creation panicked");
  }

  pushBytes(bytes: Buffer): void {
    check(container.mp3DemuxerPushBytes(this.handle, bytes, bytes.length));
  }

  /** Streams discovered so far — 0 or 1 (MP3 carries a single implicit stream). */
  streams(): Mp3StreamInfo[] {
    const count = container.mp3DemuxerStreamCount(this.handle);
    const out: Mp3StreamInfo[] = [];
    for (let i = 0; i < count; i++) {
      const raw = {} as RawStreamInfo;
      check(container.mp3DemuxerStreamAt(this.handle, i, raw));
      const extraData = copyBytes(raw.extra_data, raw.extra_data_len);
      container.streamInfoFree(raw);
      out.push({ sampleRate: raw.sample_rate, channels: raw.channels, extraData });
    }
    return out;
  }

  /** pts/duration are synthesized from a running samples-per-frame count —
   * MPEG audio carries no per-frame timing of its own. */
  pollPacket(): Packet | null {
    const raw = {} as RawPacket;
    const has: [boolean] = [false];
    check(container.mp3DemuxerPollPacket(this.handle, raw, has));
    if (!has[0]) return null;
    const payload = copyBytes(raw.payload, raw.payload_len);
    container.packetFree(raw);
    return {
      trackIndex: raw.stream_id,
      data: payload,
      pts: Number(raw.pts),
      duration: Number(raw.duration),
      key: raw.is_keyframe,
    };
  }

  close(): void {
    if (this.handle) {
      container.mp3DemuxerClose(this.handle);
      this.handle = null;
    }
  }
}
